core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -282,6 +317,154 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+    // Setup some defaults (nothing is blacklisted), then override them with particulars.
+    // (Later stubs take precedence over earlier ones.)
+    stageToMockTaskSetBlacklist.values.foreach { taskSetBlacklist =>
+      when(taskSetBlacklist.isNodeBlacklistedForTaskSet(anyString())).thenReturn(false)
+      when(taskSetBlacklist.isExecutorBlacklistedForTaskSet(anyString())).thenReturn(false)
+      when(taskSetBlacklist.isExecutorBlacklistedForTask(anyString(), anyInt())).thenReturn(false)
+      when(taskSetBlacklist.isNodeBlacklistedForTask(anyString(), anyInt())).thenReturn(false)
+    }
+    when(stageToMockTaskSetBlacklist(0).isNodeBlacklistedForTaskSet("host1")).thenReturn(true)
+    when(stageToMockTaskSetBlacklist(1).isExecutorBlacklistedForTaskSet("executor3"))
+      .thenReturn(true)
+    when(stageToMockTaskSetBlacklist(0).isExecutorBlacklistedForTask("executor0", 0))
+      .thenReturn(true)
+    IndexedSeq(
+      new WorkerOffer("executor0", "host0", 1),
+      new WorkerOffer("executor1", "host1", 1),
+      new WorkerOffer("executor2", "host1", 1),
+      new WorkerOffer("executor3", "host2", 10)
+    )
+  }
+
+  test("scheduled tasks obey task and stage blacklists") {
+    taskScheduler = setupSchedulerWithMockTaskSetBlacklist()
+    val firstTaskAttempts = taskScheduler.resourceOffers(taskSetBlacklistSetup()).flatten
+    // Whenever we schedule a task, we must consult the node and executor blacklist.  (The test
+    // doesn't check exactly what checks are made the offers get shuffled.)
+    (0 to 2).foreach { stageId =>
+      verify(stageToMockTaskSetBlacklist(stageId), atLeast(1))
+        .isNodeBlacklistedForTaskSet(anyString())
+      verify(stageToMockTaskSetBlacklist(stageId), atLeast(1))
+        .isExecutorBlacklistedForTaskSet(anyString())
+    }
+
+    // When an executor or node is blacklisted, we want to make sure that we don't try scheduling
+    // each pending task, one by one, to discover they are all blacklisted.  This is important for
+    // performance -- if we did check each task one-by-one, then responding to a resource offer
+    // (which is usually O(1)-ish) would become O(numPendingTasks), which would slow down
+    // scheduler throughput and slow down scheduling even on healthy executors.
+    // Here, we check a proxy for the runtime -- we make sure the scheduling is short-circuited
+    // at the node or executor blacklist, so we never check the per-task blacklist.
+    for {
+      exec <- Seq("executor1", "executor2")
+      part <- 0 to 1
+    } {
+      // stage 0 is blacklisted on the host for these executors
+      verify(stageToMockTaskSetBlacklist(0), never).isExecutorBlacklistedForTask(exec, part)
+    }
+    // stage 0 is blacklisted for executor3
+    (0 to 1).foreach { part =>
+      verify(stageToMockTaskSetBlacklist(1), never).isExecutorBlacklistedForTask("executor3", part)
+    }
+
+    // We should schedule all tasks.
+    assert(firstTaskAttempts.size === 6)
```

Comment:
maybe move this to right after firstTaskAttempts is assigned?  Seems better to do this check right away, since if it fails, many other things will fail, but it will be less obvious what's wrong.